File: examples/webrtc-client.js
/**
 * WebRTC SDP Client Library for RustPBX
 * 
 * This library provides a simple interface for WebRTC SDP offer/answer exchange
 * with the RustPBX server.
 * 
 * Usage:
 *   const client = new WebRTCSDPClient('http://localhost:8080');
 *   const { sessionId, answerSdp } = await client.sendOffer(offerSdp);
 */

class WebRTCSDPClient {
    /**
     * Create a new WebRTC SDP client
     * @param {string} baseUrl - Base URL of the RustPBX server
     * @param {object} options - Optional configuration
     */
    constructor(baseUrl, options = {}) {
        this.baseUrl = baseUrl.replace(/\/$/, ''); // Remove trailing slash
        this.timeout = options.timeout || 30000; // 30 second default timeout
        this.headers = {
            'Content-Type': 'application/json',
            ...options.headers
        };
    }

    /**
     * Send SDP offer and receive SDP answer
     * @param {string} offerSdp - SDP offer string
     * @param {string} sessionId - Optional session ID
     * @param {object} metadata - Optional metadata
     * @returns {Promise<{sessionId: string, answerSdp: string, iceCanvidates?: Array}>}
     */
    async sendOffer(offerSdp, sessionId = null, metadata = null) {
        const request = {
            sdp: {
                type: 'offer',
                sdp: offerSdp
            },
            ...(sessionId && { session_id: sessionId }),
            ...(metadata && { metadata })
        };

        const response = await this._makeRequest('POST', '/webrtc/offer', request);
        
        return {
            sessionId: response.session_id,
            answerSdp: response.sdp.sdp,
            iceCandidates: response.ice_candidates || []
        };
    }

    /**
     * Send ICE candidate
     * @param {string} sessionId - Session ID
     * @param {object} candidate - ICE candidate object
     * @returns {Promise<{sessionId: string, status: string}>}
     */
    async sendIceCandidate(sessionId, candidate) {
        const request = {
            session_id: sessionId,
            candidate: {
                candidate: candidate.candidate,
                sdpMLineIndex: candidate.sdpMLineIndex,
                sdpMid: candidate.sdpMid
            }
        };

        return await this._makeRequest('POST', '/webrtc/ice-candidate', request);
    }

    /**
     * Close WebRTC session
     * @param {string} sessionId - Session ID
     * @param {string} reason - Optional reason for closing
     * @returns {Promise<{sessionId: string, status: string}>}
     */
    async closeSession(sessionId, reason = null) {
        const request = {
            session_id: sessionId,
            ...(reason && { reason })
        };

        return await this._makeRequest('POST', '/webrtc/close', request);
    }

    /**
     * Get ICE servers
     * @returns {Promise<Array<{urls: Array<string>, username?: string, credential?: string}>>}
     */
    async getIceServers() {
        return await this._makeRequest('GET', '/iceservers');
    }

    /**
     * Simple SDP exchange (convenience method)
     * @param {string} offerSdp - SDP offer string
     * @returns {Promise<{sessionId: string, answerSdp: string}>}
     */
    async exchangeSdp(offerSdp) {
        const result = await this.sendOffer(offerSdp);
        return {
            sessionId: result.sessionId,
            answerSdp: result.answerSdp
        };
    }

    /**
     * Complete session setup with ICE candidates
     * @param {string} offerSdp - SDP offer string
     * @param {Array} iceCandidates - Array of ICE candidates
     * @returns {Promise<{sessionId: string, answerSdp: string}>}
     */
    async setupSession(offerSdp, iceCandidates = []) {
        // Send offer
        const result = await this.sendOffer(offerSdp);
        
        // Send ICE candidates
        for (const candidate of iceCandidates) {
            try {
                await this.sendIceCandidate(result.sessionId, candidate);
            } catch (error) {
                console.warn('Failed to send ICE candidate:', error);
            }
        }
        
        return {
            sessionId: result.sessionId,
            answerSdp: result.answerSdp
        };
    }

    /**
     * Make HTTP request to the server
     * @private
     * @param {string} method - HTTP method
     * @param {string} endpoint - API endpoint
     * @param {object} data - Request data (for POST requests)
     * @returns {Promise<object>}
     */
    async _makeRequest(method, endpoint, data = null) {
        const url = `${this.baseUrl}${endpoint}`;
        
        const options = {
            method,
            headers: this.headers,
            signal: AbortSignal.timeout(this.timeout)
        };

        if (data && (method === 'POST' || method === 'PUT')) {
            options.body = JSON.stringify(data);
        }

        try {
            const response = await fetch(url, options);
            const responseData = await response.json();

            if (!response.ok) {
                const error = new Error(responseData.error || `HTTP ${response.status}`);
                error.code = responseData.code || response.status;
                error.sessionId = responseData.session_id;
                throw error;
            }

            return responseData;
        } catch (error) {
            if (error.name === 'AbortError') {
                throw new Error('Request timeout');
            }
            throw error;
        }
    }
}

// Example usage functions
class WebRTCSDPExamples {
    constructor(client) {
        this.client = client;
    }

    /**
     * Basic WebRTC peer connection with SDP exchange
     */
    async basicPeerConnection() {
        console.log('=== Basic Peer Connection Example ===');
        
        // Create RTCPeerConnection
        const peerConnection = new RTCPeerConnection({
            iceServers: await this.client.getIceServers()
        });

        // Add audio track (you can replace this with your own media)
        const stream = await navigator.mediaDevices.getUserMedia({ 
            audio: true, 
            video: false 
        });
        
        for (const track of stream.getTracks()) {
            peerConnection.addTrack(track, stream);
        }

        // Create offer
        const offer = await peerConnection.createOffer();
        await peerConnection.setLocalDescription(offer);

        console.log('Created SDP offer');

        // Send offer to server and get answer
        const { sessionId, answerSdp } = await this.client.exchangeSdp(offer.sdp);
        
        console.log('Received SDP answer, session ID:', sessionId);

        // Set remote description
        await peerConnection.setRemoteDescription({
            type: 'answer',
            sdp: answerSdp
        });

        console.log('WebRTC connection established');

        // Return connection info for cleanup
        return { peerConnection, sessionId, stream };
    }

    /**
     * Advanced WebRTC setup with ICE candidate handling
     */
    async advancedPeerConnection() {
        console.log('=== Advanced Peer Connection Example ===');

        const iceServers = await this.client.getIceServers();
        const peerConnection = new RTCPeerConnection({ iceServers });

        // Store ICE candidates to send later
        const iceCandidates = [];
        
        peerConnection.onicecandidate = (event) => {
            if (event.candidate) {
                iceCandidates.push({
                    candidate: event.candidate.candidate,
                    sdpMLineIndex: event.candidate.sdpMLineIndex,
                    sdpMid: event.candidate.sdpMid
                });
                console.log('ICE candidate generated:', event.candidate.candidate);
            }
        };

        // Add media stream
        const stream = await navigator.mediaDevices.getUserMedia({ 
            audio: true, 
            video: false 
        });
        
        for (const track of stream.getTracks()) {
            peerConnection.addTrack(track, stream);
        }

        // Create and set local description
        const offer = await peerConnection.createOffer();
        await peerConnection.setLocalDescription(offer);

        // Wait for ICE gathering to complete
        await new Promise((resolve) => {
            if (peerConnection.iceGatheringState === 'complete') {
                resolve();
            } else {
                peerConnection.addEventListener('icegatheringstatechange', () => {
                    if (peerConnection.iceGatheringState === 'complete') {
                        resolve();
                    }
                });
            }
        });

        console.log('ICE gathering complete, collected', iceCandidates.length, 'candidates');

        // Setup complete session
        const { sessionId, answerSdp } = await this.client.setupSession(offer.sdp, iceCandidates);

        console.log('Session setup complete, session ID:', sessionId);

        // Set remote description
        await peerConnection.setRemoteDescription({
            type: 'answer',
            sdp: answerSdp
        });

        console.log('Advanced WebRTC connection established');

        return { peerConnection, sessionId, stream };
    }

    /**
     * Example with metadata and error handling
     */
    async metadataExample() {
        console.log('=== Metadata Example ===');

        try {
            const peerConnection = new RTCPeerConnection({
                iceServers: await this.client.getIceServers()
            });

            // Add data channel
            const dataChannel = peerConnection.createDataChannel('chat', {
                ordered: true
            });

            // Create offer
            const offer = await peerConnection.createOffer();
            await peerConnection.setLocalDescription(offer);

            // Send offer with metadata
            const metadata = {
                userId: 'user123',
                callType: 'data',
                timestamp: new Date().toISOString(),
                features: ['chat', 'file-transfer']
            };

            const result = await this.client.sendOffer(
                offer.sdp,
                'metadata-session-' + Date.now(),
                metadata
            );

            console.log('Session created with metadata:', result.sessionId);

            // Set remote description
            await peerConnection.setRemoteDescription({
                type: 'answer',
                sdp: result.answerSdp
            });

            return { peerConnection, sessionId: result.sessionId, dataChannel };

        } catch (error) {
            console.error('Metadata example failed:', error.message);
            if (error.code) {
                console.error('Error code:', error.code);
            }
            throw error;
        }
    }

    /**
     * Concurrent sessions example
     */
    async concurrentSessions() {
        console.log('=== Concurrent Sessions Example ===');

        const sessions = [];
        const promises = [];

        for (let i = 0; i < 3; i++) {
            const promise = this.createSession(`concurrent-${i}`);
            promises.push(promise);
        }

        const results = await Promise.allSettled(promises);
        
        for (let i = 0; i < results.length; i++) {
            if (results[i].status === 'fulfilled') {
                sessions.push(results[i].value);
                console.log(`Session ${i} created successfully:`, results[i].value.sessionId);
            } else {
                console.error(`Session ${i} failed:`, results[i].reason.message);
            }
        }

        return sessions;
    }

    /**
     * Helper function to create a session
     * @private
     */
    async createSession(prefix) {
        const peerConnection = new RTCPeerConnection({
            iceServers: await this.client.getIceServers()
        });

        // Add a dummy audio track
        const audioContext = new AudioContext();
        const oscillator = audioContext.createOscillator();
        const destination = audioContext.createMediaStreamDestination();
        oscillator.connect(destination);
        oscillator.start();

        const track = destination.stream.getAudioTracks()[0];
        peerConnection.addTrack(track, destination.stream);

        const offer = await peerConnection.createOffer();
        await peerConnection.setLocalDescription(offer);

        const { sessionId, answerSdp } = await this.client.exchangeSdp(offer.sdp);

        await peerConnection.setRemoteDescription({
            type: 'answer',
            sdp: answerSdp
        });

        return { peerConnection, sessionId, audioContext };
    }

    /**
     * Cleanup resources
     */
    async cleanup(connections) {
        for (const conn of connections) {
            if (conn.peerConnection) {
                conn.peerConnection.close();
            }
            if (conn.stream) {
                conn.stream.getTracks().forEach(track => track.stop());
            }
            if (conn.audioContext) {
                await conn.audioContext.close();
            }
            if (conn.sessionId) {
                try {
                    await this.client.closeSession(conn.sessionId, 'Example cleanup');
                } catch (error) {
                    console.warn('Failed to close session:', error.message);
                }
            }
        }
    }
}

// Export for use in Node.js or browsers
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { WebRTCSDPClient, WebRTCSDPExamples };
} else if (typeof window !== 'undefined') {
    window.WebRTCSDPClient = WebRTCSDPClient;
    window.WebRTCSDPExamples = WebRTCSDPExamples;
}

// Example usage (for browser console or Node.js)
async function runExamples() {
    console.log('Starting WebRTC SDP Client Examples');
    
    const client = new WebRTCSDPClient('http://localhost:8080');
    const examples = new WebRTCSDPExamples(client);
    
    const connections = [];
    
    try {
        // Run basic example
        const basic = await examples.basicPeerConnection();
        connections.push(basic);
        
        // Wait a bit
        await new Promise(resolve => setTimeout(resolve, 2000));
        
        // Run advanced example
        const advanced = await examples.advancedPeerConnection();
        connections.push(advanced);
        
        // Wait a bit
        await new Promise(resolve => setTimeout(resolve, 2000));
        
        // Run metadata example
        const metadata = await examples.metadataExample();
        connections.push(metadata);
        
        // Wait a bit
        await new Promise(resolve => setTimeout(resolve, 2000));
        
        // Run concurrent sessions
        const concurrent = await examples.concurrentSessions();
        connections.push(...concurrent);
        
        console.log('All examples completed successfully!');
        
        // Wait before cleanup
        await new Promise(resolve => setTimeout(resolve, 5000));
        
    } catch (error) {
        console.error('Example failed:', error);
    } finally {
        // Cleanup
        await examples.cleanup(connections);
        console.log('Cleanup completed');
    }
}

// Uncomment to run examples automatically
// runExamples().catch(console.error);

File: src/webrtc_client.rs
use anyhow::Result;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::time::Duration;
use tracing::{error, info};

/// WebRTC SDP Session Description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SdpSessionDescription {
    #[serde(rename = "type")]
    pub sdp_type: String,
    pub sdp: String,
}

/// WebRTC ICE Candidate
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IceCandidate {
    pub candidate: String,
    #[serde(rename = "sdpMLineIndex")]
    pub sdp_m_line_index: Option<u16>,
    #[serde(rename = "sdpMid")]
    pub sdp_mid: Option<String>,
}

/// SDP Offer Request
#[derive(Debug, Serialize)]
pub struct SdpOfferRequest {
    pub sdp: SdpSessionDescription,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Value>,
}

/// SDP Answer Response
#[derive(Debug, Deserialize)]
pub struct SdpAnswerResponse {
    pub sdp: SdpSessionDescription,
    pub session_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ice_candidates: Option<Vec<IceCandidate>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Value>,
}

/// ICE Candidate Request
#[derive(Debug, Serialize)]
pub struct IceCandidateRequest {
    pub session_id: String,
    pub candidate: IceCandidate,
}

/// ICE Candidate Response
#[derive(Debug, Deserialize)]
pub struct IceCandidateResponse {
    pub session_id: String,
    pub status: String,
}

/// Close Session Request
#[derive(Debug, Serialize)]
pub struct CloseSessionRequest {
    pub session_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

/// Close Session Response
#[derive(Debug, Deserialize)]
pub struct CloseSessionResponse {
    pub session_id: String,
    pub status: String,
}

/// Error Response
#[derive(Debug, Deserialize)]
pub struct ErrorResponse {
    pub error: String,
    pub code: u16,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
}

/// WebRTC Client for SDP offer/answer exchange
pub struct WebRtcClient {
    client: Client,
    base_url: String,
}

impl WebRtcClient {
    /// Create a new WebRTC client
    pub fn new(base_url: String) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .expect("Failed to create HTTP client");

        Self { client, base_url }
    }

    /// Create a new WebRTC client with custom HTTP client
    pub fn new_with_client(base_url: String, client: Client) -> Self {
        Self { client, base_url }
    }

    /// Send SDP offer and receive SDP answer
    pub async fn send_offer(
        &self,
        offer_sdp: String,
        session_id: Option<String>,
        metadata: Option<Value>,
    ) -> Result<SdpAnswerResponse> {
        let request = SdpOfferRequest {
            sdp: SdpSessionDescription {
                sdp_type: "offer".to_string(),
                sdp: offer_sdp,
            },
            session_id,
            metadata,
        };

        let url = format!("{}/webrtc/offer", self.base_url);
        info!("Sending SDP offer to: {}", url);

        let response = self
            .client
            .post(&url)
            .json(&request)
            .send()
            .await?;

        if response.status().is_success() {
            let answer: SdpAnswerResponse = response.json().await?;
            info!("Received SDP answer for session: {}", answer.session_id);
            Ok(answer)
        } else {
            let error: ErrorResponse = response.json().await?;
            Err(anyhow::anyhow!("Server error: {} (code: {})", error.error, error.code))
        }
    }

    /// Send ICE candidate
    pub async fn send_ice_candidate(
        &self,
        session_id: String,
        candidate: IceCandidate,
    ) -> Result<IceCandidateResponse> {
        let request = IceCandidateRequest {
            session_id: session_id.clone(),
            candidate,
        };

        let url = format!("{}/webrtc/ice-candidate", self.base_url);
        info!("Sending ICE candidate for session: {}", session_id);

        let response = self
            .client
            .post(&url)
            .json(&request)
            .send()
            .await?;

        if response.status().is_success() {
            let result: IceCandidateResponse = response.json().await?;
            Ok(result)
        } else {
            let error: ErrorResponse = response.json().await?;
            Err(anyhow::anyhow!("Server error: {} (code: {})", error.error, error.code))
        }
    }

    /// Close WebRTC session
    pub async fn close_session(
        &self,
        session_id: String,
        reason: Option<String>,
    ) -> Result<CloseSessionResponse> {
        let request = CloseSessionRequest {
            session_id: session_id.clone(),
            reason,
        };

        let url = format!("{}/webrtc/close", self.base_url);
        info!("Closing session: {}", session_id);

        let response = self
            .client
            .post(&url)
            .json(&request)
            .send()
            .await?;

        if response.status().is_success() {
            let result: CloseSessionResponse = response.json().await?;
            Ok(result)
        } else {
            let error: ErrorResponse = response.json().await?;
            Err(anyhow::anyhow!("Server error: {} (code: {})", error.error, error.code))
        }
    }

    /// Get ICE servers
    pub async fn get_ice_servers(&self) -> Result<Vec<IceServer>> {
        let url = format!("{}/iceservers", self.base_url);
        info!("Getting ICE servers from: {}", url);

        let response = self.client.get(&url).send().await?;

        if response.status().is_success() {
            let ice_servers: Vec<IceServer> = response.json().await?;
            Ok(ice_servers)
        } else {
            Err(anyhow::anyhow!("Failed to get ICE servers: HTTP {}", response.status()))
        }
    }
}

/// ICE Server configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IceServer {
    pub urls: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credential: Option<String>,
}

/// Convenience functions for common WebRTC operations
impl WebRtcClient {
    /// Simple offer/answer exchange
    pub async fn exchange_sdp(
        &self,
        offer_sdp: String,
    ) -> Result<(String, String)> {
        let answer_response = self.send_offer(offer_sdp, None, None).await?;
        Ok((answer_response.session_id, answer_response.sdp.sdp))
    }

    /// Offer/answer exchange with session tracking
    pub async fn exchange_sdp_with_session(
        &self,
        offer_sdp: String,
        session_id: String,
    ) -> Result<String> {
        let answer_response = self.send_offer(offer_sdp, Some(session_id), None).await?;
        Ok(answer_response.sdp.sdp)
    }

    /// Complete WebRTC session setup
    pub async fn setup_session(
        &self,
        offer_sdp: String,
        ice_candidates: Vec<IceCandidate>,
    ) -> Result<(String, String)> {
        // Send offer and get answer
        let answer_response = self.send_offer(offer_sdp, None, None).await?;
        let session_id = answer_response.session_id.clone();
        
        // Send ICE candidates
        for candidate in ice_candidates {
            if let Err(e) = self.send_ice_candidate(session_id.clone(), candidate).await {
                error!("Failed to send ICE candidate: {}", e);
            }
        }
        
        Ok((session_id, answer_response.sdp.sdp))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use mockall::predicate::*;
    use serde_json::json;

    #[tokio::test]
    async fn test_webrtc_client_creation() {
        let client = WebRtcClient::new("http://localhost:8080".to_string());
        assert_eq!(client.base_url, "http://localhost:8080");
    }

    #[tokio::test]
    async fn test_sdp_serialization() {
        let offer_request = SdpOfferRequest {
            sdp: SdpSessionDescription {
                sdp_type: "offer".to_string(),
                sdp: "v=0\r\no=- 123456789 123456789 IN IP4 192.168.1.1\r\n".to_string(),
            },
            session_id: Some("test-session".to_string()),
            metadata: Some(json!({"test": "data"})),
        };

        let json_str = serde_json::to_string(&offer_request).unwrap();
        assert!(json_str.contains("offer"));
        assert!(json_str.contains("test-session"));
    }

    #[tokio::test]
    async fn test_ice_candidate_serialization() {
        let candidate = IceCandidate {
            candidate: "candidate:1 1 UDP 2013266431 192.168.1.1 54400 typ host".to_string(),
            sdp_m_line_index: Some(0),
            sdp_mid: Some("0".to_string()),
        };

        let json_str = serde_json::to_string(&candidate).unwrap();
        assert!(json_str.contains("candidate:1"));
        assert!(json_str.contains("sdpMLineIndex"));
    }
}